src/sync.js

```diff
@@ -352,77 +350,88 @@ class Sync {
   // If a file has been changed, we had to check what operation it is.
   // For a move, the first call will just keep a reference to the document,
   // and only at the second call, the move operation will be executed.
-  fileChanged (doc: Metadata, side: Side, rev: number, callback: Callback) {
+  async fileChangedAsync (doc: Metadata, side: Side, rev: number): Promise<void> {
     let from
     switch (true) {
       case doc._deleted && (rev === 0):
-        callback()
-        break
+        return
       case this.moveFrom != null:
         // $FlowFixMe
         from = (this.moveFrom: Metadata)
         this.moveFrom = null
         if (from.moveTo === doc._id) {
-          side.moveFile(doc, from, err => {
-            if (err) { this.moveFrom = from }
-            callback(err)
-          })
+          try {
+            await side.moveFileAsync(doc, from)
+          } catch (err) {
+            this.moveFrom = from
+            throw err
+          }
         } else {
           log.error('Invalid move')
           log.error(from)
           log.error(doc)
-          side.addFile(doc, function (err) {
-            if (err) { log.error(err) }
-            side.destroy(from, function (err) {
-              if (err) { log.error(err) }
-              callback(new Error('Invalid move'))
-            })
-          })
+          try {
+            await side.addFileAsync(doc)
+          } catch (err) {
+            log.error(err)
+            try {
+              await side.destroyAsync(from)
+            } catch (err) {
+              log.error(err)
+              throw new Error('Invalid move')
+            }
+          }
         }
         break
       case doc.moveTo != null:
         this.moveFrom = doc
-        callback()
-        break
+        return
       case doc._deleted:
-        side.destroy(doc, callback)
-        break
+        return side.destroyAsync(doc)
       case rev === 0:
-        side.addFile(doc, callback)
-        break
+        return side.addFileAsync(doc)
       default:
-        this.pouch.getPreviousRev(doc._id, rev, function (err, old) {
-          if (err) {
-            side.overwriteFile(doc, old, callback)
-          } else if (old.checksum === doc.checksum) {
-            side.updateFileMetadata(doc, old, callback)
-          } else if (old.remote && !old.checksum) {
-            // Photos uploaded by cozy-mobile have no checksum,
-            // but it's useless to reupload the binary
-            side.updateFileMetadata(doc, old, callback)
-          } else {
-            side.overwriteFile(doc, old, callback)
-          }
-        })
+        let old
+        try {
+          old = this.pouch.getPreviousRevAsync(doc._id, rev)
+        } catch (err) {
+          // FIXME: overwriteFileAsync never uses old anyway
+          return side.overwriteFileAsync(doc, doc)
```

Comment:
Old line 396 bothered me... Better explicitely pass `null` instead of `old`, or `doc` (which is worse)...
I'll need to fix `overwriteFileAsync()`